doc/debugger.rst

```diff
@@ -1773,6 +1811,161 @@ target endianness and interpret the bytecode based on that.
+        
+        if (strcmp(cmdname, "VersionInfo") == 0) {
+            /* Return a positive integer to send a REP containing values pushed
+             * to the stack. The return value indicates how many dvalues you
+             * are including in the response.
+             */
+            
+            duk_push_string(ctx, "My Awesome Program");
+            duk_push_int(ctx, 81200);  /* ver. 8.12.0 */
+            
+            return 2;  /* 2 dvalues */
+        } else {
+            /* Return -1 to send an ERR reply. The value on top of the stack
+             * will be string coerced and used for the error text.
+             */
+            
+            duk_push_string(ctx, "unrecognized AppRequest command name");
+            return -1;
+        }
+    }
+
+If no request callback is provided at attach, AppRequest will be treated as an
+unsupported command, eliciting an ERR reply from Duktape saying so.  Such a
+target is still free to send AppNotify messages, however.
+
+As a precaution, the target should try to avoid sending structured values such
+as JS objects in notify messages as their heap pointers may become stale by the
+time the client receives them.  This is especially true for notifications sent
+while the target is running.  It's better to stick to primitives which have
+unique dvalue representations, i.e. numbers, strings and buffers.  If a
+structured value does need to be sent, it can simply be, e.g. JSON/JX encoded
+and sent as a string instead.
+
+Important notes on the request callback
+---------------------------------------
+
+The request callback is provided with a ``duk_context`` pointer with which it
+can access the value stack and is assumed to be trusted.  Specifically, there
+are certain things it **MUST NOT** do.  Specifically:
+
+* It MUST NOT attempt to access or pop any values from the top of the stack
+  beyond the ``nvalues`` it is given.
+
+* It MUST NOT assume any specific value for ``duk_get_top()`` and similar
+  primitives.  In practice this means using negative stack indices to access
+  values.
+
+* It MUST NOT throw errors.  It is very easy to accidentally throw an error
+  when working with value stack values directly, so caution must be exercised
+  here.
+
+Violating this contract is undefined behavior and may corrupt debugger state,
+cause incorrect behavior, or even lead to a segfault.  In the future it would
+be nice to make this more robust, e.g. by sandboxing the function so that it
+cannot access unrelated stack values.
+
+The dvalues of a message are pushed in the order they are received.  This makes
+them inconvenient to access using negative indices, since the relative position
+of any given value on the stack is dependent on the total number of values.
+However because the callback receives the total number of values as a parameter,
+a useful convention is to index the stack like so::
+
+    cmd_name = duk_get_string(ctx, -nvalues + 0);
```

Comment:
I don't think it's obvious because in a normal Duktape/C function it's safe to for example `duk_get_string()` for an invalid index: if the index doesn't exist you get a NULL and can resume predictable from there. But here you might be reading an unrelated value as a string (and if you're unlucky it might actually be a string).